internal/providers/github/factory/factory.go

```diff
@@ -0,0 +1,223 @@
+// Copyright 2024 Stacklok, Inc.
```

Comment:
The code here is mostly copy-paste from `ProviderBuilder`. The main difference is that the process of figuring out which type of credential is determined dynamically here, whereas `ProviderBuilder` requires it to be figured out when the `ProviderBuilder` struct is instantiated.

Reply:
Another note: I have decided to rely on existing test code to verify this logic once it is wired up in the application.